Improve/remove PATH environment modifications

terminalmage opened the pull request:
- `salt/modules/rbenv.py`: Forces use of str types in the custom env dict passed to `cmd.run_all`.
- `salt/modules/syslog_ng.py`: Ditches janky PATH munging in favor of the stable and long-existing support built into cmdmod.py
- `salt/modules/win_path.py`: Forces use of str types in path modification functions.
- `salt/states/win_path.py`: Completely rewritten. Duplicated code from the execution module removed in favor of calls to the execution module. Tests junked and 14 new tests written.
- `salt/utils/path.py`: The `which()` function was mistakenly modified in 20033ee to inject the directories in the POSIX default path into the PATH environment variable (even for Windows!). We never used the PATH to find the executables, we simply cycled through the dirs one by one and looked for an executable file matching the named path. The code that modifies the path is now removed. In addition, `which()` now uses `salt.utils.path.join()`, which gracefully handles mismatched str and unicode directory components to prevent decode errors.  `join()` has also been simplified to use `salt.utils.data.decode()` to normalize directory components to unicode.

terminalmage committed:
Improve/remove PATH environment modifications

- `salt/modules/rbenv.py`: Forces use of str types in the custom env dict
  passed to `cmd.run_all`.
- `salt/modules/syslog_ng.py`: Ditches janky PATH munging in favor of
  the stable and long-existing support built into cmdmod.py
- `salt/modules/win_path.py`: Forces use of str types in path
  modification functions.
- `salt/states/win_path.py`: Completely rewritten. Duplicated code from
  the execution module removed in favor of calls to the execution
  module. Tests junked and 14 new tests written.
- `salt/utils/path.py`: The `which()` function was mistakenly modified
  in 20033ee to inject the directories in the POSIX default path into
  the PATH environment variable (even for Windows!). We never used the
  PATH to find the executables, we simply cycled through the dirs one by
  one and looked for an executable file matching the named path. The
  code that modifies the path is now removed. In addition, `which()` now
  uses `salt.utils.path.join()`, which gracefully handles mismatched
  str and unicode directory components to prevent decode errors.
  `join()` has also been simplified to use `salt.utils.data.decode()` to
  normalize directory components to unicode.

terminalmage commented on salt/modules/rbenv.py:
```diff
@@ -380,7 +380,12 @@ def do(cmdline, runas=None, env=None):
+            path,
+            salt.utils.stringutils.to_unicode(os.environ['PATH'])
+        )
+    )
```
NOTE: I know it may seem weird that this code converts the PATH env var to unicode and then back to a `str` type. The unicode conversion is to ensure that the `.format()` doesn't raise a decode error, and the `to_str()` ensures the result is a `str` type like we need it to be.

rallytime replied:
Maybe you should add a comment above explaining this, just so someone doesn't see it in the code and remove it in the future.

damon-atkins commented on salt/modules/win_path.py:
```diff
@@ -126,10 +130,16 @@ def add(path, index=0):
         index = len(sysPath)
 
-    localPath = os.environ["PATH"].split(os.pathsep)
+    pathsep = str(os.pathsep)  # future lint: disable=blacklisted-function
```
Windows registry is Unicode and the PATH should be Unicode as well. As NTFS is Unicode filenames.
FAT based files systems do not support Unicode, but that would be unto the person to make sure the input is valid ASCII.
So the reg function read the current path will return Unicode, and you need to send Unicode back when setting it.

terminalmage replied:
It seems that `reg.set_value` would convert to unicode anyway, so I'm not sure that's too big of a concern, but I've changed the calls to it so that a unicode PATH string is passed to it.

You are wrong however, with respect to `os.environ` containing unicode data. On Python 2 this causes a traceback:

```
  (snip)
  ...
  File "c:\python27\lib\subprocess.py", line 390, in __init__
    errread, errwrite)
  File "c:\python27\lib\subprocess.py", line 640, in _execute_child
    startupinfo)
TypeError: environment can only contain strings
```

damon-atkins replied:
Thanks for trying.  It just provides more proof that PY2 is no longer suitable for Windows, PY2 has too many hacks to work in modern versions of Windows which use Unicode.  Even PowerShell 6 has now moved to UTF-8 without BOM instead of UTF with BOM to be more cross platform friendly.

rallytime commented on salt/modules/syslog_ng.py:
```diff
@@ -784,49 +784,27 @@ def set_parameters(version=None,
+                      syslog_ng_sbin_dir)
+        else:
+            env = {
+                str('PATH'): salt.utils.stringutils.to_str(  # future lint: disable=blacklisted-function
```
A comment about this `str()` and `to_str()` behavior would likely be good here as well.

rallytime reviewed (changes requested):
Overall, this is really well done. I have a couple of small comments.
This needs to be rebased and it also looks like this is causing the `unit.modules.test_syslog_ng` tests to fail.

I'd like to see what @twangboy finds during his testing as well.

terminalmage committed:
Add comment to explain PATH handling

Also use salt.utils.path.join()

terminalmage committed:
Pass unicode path to reg.set_value

terminalmage commented:
@rallytime I still need to address your concerns, will do so shortly.

terminalmage committed:
Add comment regarding PATH handling

terminalmage commented:
@rallytime ready for re-review

terminalmage committed:
Add support for negative indexes

terminalmage commented:
I've just added support for negative indexes to this PR, and added new tests.

terminalmage committed:
win_path: Do not normalize path to lowercase

This removes the lowercase normalization and handles case-insensitive
path matches. It also combines local path and system path inspection
into a single helper function in both add() and remove(), for simplicity.

Additionally, the add/remove unit tests have been rewritten to test
a wider array of use cases.

terminalmage commented:
@rallytime @twangboy please re-review, I've updated the state and execution module to properly handle case-insensitive paths without converting them to lowercase (see https://github.com/saltstack/salt/issues/45409).

rallytime reviewed (approved):
Nicely done, and thanks for adding tests!

damon-atkins commented on salt/modules/win_path.py:
```diff
@@ -68,9 +78,12 @@ def get_path():
-    ret = __salt__['reg.read_value']('HKEY_LOCAL_MACHINE',
-                                   'SYSTEM\\CurrentControlSet\\Control\\Session Manager\\Environment',
-                                   'PATH')['vdata'].split(';')
+    ret = salt.utils.stringutils.to_unicode(
```
__salt__['reg.*'] always returns Unicode since 2016.11, if it does not its a bug.

rallytime commented:
@terminalmage I think the syslog_ng test failures are related here. See `unit.modules.test_syslog_ng` on Python 3 tests.

terminalmage committed:
Update syslog_ng tests to reflect change in PATH behavior

rallytime commented:
@terminalmage There are some lint errors in the test file. Can you take a look?

rallytime commented:
@terminalmage A couple of related failures still: 
- unit.modules.test_syslog_ng.SyslogNGTestCase.test_end_to_end_statement_generation
- unit.modules.test_syslog_ng.SyslogNGTestCase.test_option_with_parameters
- unit.modules.test_syslog_ng.SyslogNGTestCase.test_value_with_arguments

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/19528/#showFailuresLink

rallytime commented:
@terminalmage There's a lint error on this PR too: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/18938/violations/

I think some of the failures here are related as well: https://jenkins.saltstack.com/job/PR/job/salt-pr-rs-cent7-n/15983/#showFailuresLink

Can you take a look?

terminalmage commented:
Lint error fixed, and the Cent7 failures were due to an unrelated issue that was fixed by @garethgreenaway in a separate PR. I've rebased, they should pass now.
